design-documents/bit-precise-types.rst

```diff
@@ -0,0 +1,213 @@
+  that programmers may use have a more familiar representation.
+
+- Operations ``*,/``, saving and loading values to memory, and casting to
+  another type would all require extra cost.
+
+- Operations ``+,-`` on "large" values (greater than one register) would require
+  an extra instruction to "normalize" the carry-bit.
+
+
+Option ``B`` has the following benefits:
+
+- For small values in memory, the AArch64 ``LDADD`` operations work naturally.
+
+- Operations ``+,-,*,<<``, narrowing conversions, and loading/storing to memory
+  would all naturally work.
+
+- On AArch64 this would most likely match the expectation of developers, and
+  e.g. a ``_BitInt(8)`` would have the same representation as a ``char`` in
+  registers.
+
+It has the following negatives:
+
+- The AArch64 ``LD{S,U}MAX`` operations would not work naturally on small values
+  of this representation.
+
+- Operations ``/,%,==,<,>,<=,>=,>>`` and widening conversions would not
+  require extra work.
+
+- On AArch32 this could cause surprises to developers, given that on this
+  architecture small Fundamental Data Types are have zero- or sign-extended
+  extra bits.  So a ``char`` would not have the same representation as a
+  ``_BitInt(8)`` on this architecture.
+
+
+Option ``C`` has the following benefits:
+
+- For small values in memory, the AArch64 ``LD{S,U}MAX`` operations work
+  naturally.
+
+- Operations ``==,<,<=,>=,>,>>``, widening conversions, and loading/storing to
+  memory would all naturally work.
+
+- On AArch32 this could match the expectation of developers, with a
+  ``_BitInt(8)`` in a register matching the representation of a ``char``.
+
+It has the following negatives:
+
+- The AArch64 ``LDADD`` operations would not work naturally.
+
+- Operations ``+,-,*,<<`` would all cause the need for masking at an ABI
+  boundary.
+
+- On AArch64 this would not match the expectation of developers, with
+  ``_BitInt(8)`` not matching the representation of a ``char``.
+
+Summary, suggestion, and reasoning
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Overall it seems that for operations on small values option ``A`` is more
+performant.  However, when acting on "large" values (i.e. greater than the size
+of one register) it loses some of that benefit.  Storing to and from memory
+would also come at a cost for this representation.  This is also likely to be
+the most surprising representation for developers on an Arm platform.
+
+Between option ``B`` and option ``C`` there is not a great difference in
+performance characteristics.  However it should be noted that option ``C`` is
+likely the most natural extension of the AArch32 PCS rules for unspecified bits
+in a register containing a small Fundamental Data Type, while option ``B`` is
+the most natural extension of the similar rules in AArch64 PCS.
+
+As mentioned above, we do not expect operations on ``_BitInt`` types to be
+performance critical.  Given that providing a productive environment for
```

Comment:
Overall, I'm not sure if we can expect _BitInt to never be used in
performance-critical code.
If people will write code targeting FPGAs with _BitInt, presumably that
code is performance-critical (why would anyone otherwise target an FPGA
for that code?). If such code becomes popular, at some point, there may
be good reasons to use the same source code to target CPUs too. In that
case, the _BitInt code would be performance critical...

Reply:
Perhaps it's more coherent to say "As mentioned above, we do not expect operations on ``_BitInt`` types _at large sizes_ to be performance critical." since the benefits of `_BitInt` allowing sidestepping the arithmetic promotions disappear for sizes larger than `_BitInt(32)`.

Reply:
Perhaps it would be more coherent to say "As mentioned above, we do not expect operations on ``_BitInt`` types _with N greater than 32 bits_ to be performance critical." due to N2763's rationale.